miner/Miner.hs

```diff
@@ -244,29 +226,60 @@ scheme env = case cmd $ args env of
     CPU e -> cpu (version $ args env) e (gen env)
     GPU _ -> gpu
 
--- | Submit a new `BlockHeader` to mine (i.e. to determine a valid `Nonce`).
+-- | Attempt to get new work while obeying a sane retry policy.
 --
-submit :: Env -> WorkBytes -> IO ()
-submit (work -> w) bs = atomically $
-    isEmptyTMVar w >>= bool (void $ swapTMVar w bs) (putTMVar w bs)
+getWork :: Env -> IO (Maybe WorkBytes)
+getWork e = retrying policy (\_ -> pure . isNothing) $ const f
+  where
+    -- | If we wait longer than the average block time and still can't get
+    -- anything, then there's no point in continuing to wait.
+    --
+    policy :: RetryPolicy
+    policy = exponentialBackoff 500000 <> limitRetries 7
```

Comment:
500ms, 1s, 2s, 4s, 8s, 16s, 32s... is seven attempts. I considered also making the `NodeUrl` flag instead accept a list, so that `getWork` can fallback to other nodes if an initial attempt fails.